numba/cuda/ufuncs.py

```diff
@@ -628,4 +637,26 @@ def np_real_atanh_impl(context, builder, sig, args):
         'qq->q': numbers.int_shr_impl,
         'QQ->Q': numbers.int_shr_impl,
     }
+
+    db[np.log] = {
+        'f->f': np_real_log_impl,
+        'd->d': np_real_log_impl,
+        'F->F': npyfuncs.np_complex_log_impl,
+        'D->D': npyfuncs.np_complex_log_impl,
```

Comment:
I'm not sure what will happen with complex data types, I think the calls will resolve in terms of the `math` and `cmath` implementations, but am not sure how they will interact with CUDA `math`.

Reply:
`np.log(complex_array) is implemented using a combination of three libdevice functions: `hypotf`, `logf` and `atan2f`:

```llvm
...
B0:
  %".25" = call float @"__nv_hypotf"(float %"arg.x", float %"arg.y")
  %".26" = call float @"__nv_logf"(float %".25")
  %".27" = call float @"__nv_atan2f"(float %"arg.y", float %"arg.x")
  store {float, float} zeroinitializer, {float, float}* %".28"
  %".31" = getelementptr inbounds {float, float}, {float, float}* %".28", i32 0, i32 0
  store float %".26", float* %".31"
  %".33" = getelementptr inbounds {float, float}, {float, float}* %".28", i32 0, i32 1
  store float %".27", float* %".33"
  %".35" = load {float, float}, {float, float}* %".28"
  %"extracted.real" = extractvalue {float, float} %".35", 0
  %"extracted.imag" = extractvalue {float, float} %".35", 1
  %".36" = insertvalue {float, float} undef, float %"extracted.real", 0
  %".37" = insertvalue {float, float} %".36", float %"extracted.imag", 1
  store {float, float} %".37", {float, float}* %".ret"
  ret i32 0
```

Reply:
These libdevice functions are used because of the lowerings of `math.log`, `math.hypot`, and `math.atan2` in [`mathimpl.py`](https://github.com/numba/numba/blob/main/numba/cuda/mathimpl.py).